File: vtk-rs/libvtkrs/src/vtk_object_base.cpp
#include "vtk_object_base.h"
#include "cxx.h"
#include "vtkIndent.h"
#include "vtk_object_base.rs.h"
#include <sstream>

#include <cstdint>
#include <vtkObjectBase.h>

rust::String vtk_object_base_get_class_name(const vtkObjectBase& obj) {
    return obj.GetClassName();
}

bool vtk_object_base_is_a(const vtkObjectBase& obj, rust::Str class_name) {
    return const_cast<vtkObjectBase&>(obj).IsA(class_name.begin());
}

std::int64_t vtk_object_base_get_number_of_generations_from_base(
    const vtkObjectBase& obj, rust::Str base_class
) {
    return obj.GetNumberOfGenerationsFromBaseType(base_class.begin());
}

void vtk_object_base_fast_delete(vtkObjectBase& obj) {
    obj.FastDelete();
}

std::int64_t vtk_object_base_get_reference_count(const vtkObjectBase& obj) {
    return const_cast<vtkObjectBase&>(obj).GetReferenceCount();
}

void vtk_object_base_set_reference_count(vtkObjectBase& obj, std::int64_t count) {
    obj.SetReferenceCount(count);
}

bool vtk_object_base_get_is_in_memkind(const vtkObjectBase& obj) {
    return obj.GetIsInMemkind();
}

rust::String vtk_object_base_print_self(const vtkObjectBase& obj, std::uint64_t indent) {
    std::stringstream oss;
    const_cast<vtkObjectBase&>(obj).PrintSelf(oss, vtkIndent(indent));
    return oss.str();
}

rust::String vtk_object_base_print_header(const vtkObjectBase& obj, std::uint64_t indent) {
    std::stringstream oss;
    const_cast<vtkObjectBase&>(obj).PrintHeader(oss, vtkIndent(indent));
    return oss.str();
}

rust::String vtk_object_base_print_trailer(const vtkObjectBase& obj, std::uint64_t indent) {
    std::stringstream oss;
    const_cast<vtkObjectBase&>(obj).PrintTrailer(oss, vtkIndent(indent));
    return oss.str();
}

rust::String vtk_object_base_get_object_description(const vtkObjectBase& obj) {
#ifdef VTK094
    return obj.GetObjectDescription();
#else
    return "";
#endif
}

bool vtk_object_base_uses_garbage_collector(const vtkObjectBase& obj) {
#ifdef VTK094
    return const_cast<vtkObjectBase&>(obj).UsesGarbageCollector();
#else
    return "";
#endif
}

File: vtk-rs/src/vtk_object_base.rs
#[cxx::bridge]
pub(crate) mod ffi {
    unsafe extern "C++" {
        include!("vtk_object_base.h");

        type vtkObjectBase;

        fn vtk_object_base_get_class_name(obj: &vtkObjectBase) -> String;
        #[allow(unused)]
        fn vtk_object_base_get_object_description(obj: &vtkObjectBase) -> String;
        fn vtk_object_base_is_a(obj: &vtkObjectBase, class_name: &str) -> bool;
        fn vtk_object_base_get_number_of_generations_from_base(
            obj: &vtkObjectBase,
            base_class: &str,
        ) -> i64;
        fn vtk_object_base_fast_delete(obj: Pin<&mut vtkObjectBase>);
        fn vtk_object_base_get_reference_count(obj: &vtkObjectBase) -> i64;
        fn vtk_object_base_set_reference_count(obj: Pin<&mut vtkObjectBase>, count: i64);
        fn vtk_object_base_get_is_in_memkind(obj: &vtkObjectBase) -> bool;
        fn vtk_object_base_print_self(obj: &vtkObjectBase, indent: u64) -> String;
        fn vtk_object_base_print_header(obj: &vtkObjectBase, indent: u64) -> String;
        fn vtk_object_base_print_trailer(obj: &vtkObjectBase, indent: u64) -> String;
        #[allow(unused)]
        fn vtk_object_base_uses_garbage_collector(obj: &vtkObjectBase) -> bool;
    }
}

pub(crate) mod private {
    pub trait Sealed {}
}

/// [`vtkObjectBase`](https://vtk.org/doc/nightly/html/classvtkObjectBase.html)
#[allow(non_camel_case_types)]
pub trait vtkObjectBase: private::Sealed {
    #[doc(hidden)]
    fn as_vtk_object_base(&self) -> core::pin::Pin<&ffi::vtkObjectBase>;
    #[doc(hidden)]
    fn as_vtk_object_base_mut(&mut self) -> core::pin::Pin<&mut ffi::vtkObjectBase>;

    #[doc(alias = "GetClassName")]
    fn get_class_name(&self) -> String {
        ffi::vtk_object_base_get_class_name(&self.as_vtk_object_base())
    }

    #[cfg(feature = "v094")]
    #[cfg_attr(docsrs, doc(cfg(feature = "v094")))]
    #[doc(alias = "GetObjectDescription")]
    fn get_object_description(&self) -> String {
        ffi::vtk_object_base_get_object_description(&self.as_vtk_object_base())
    }

    #[doc(alias = "IsA")]
    fn is_a(&self, class_name: &str) -> bool {
        ffi::vtk_object_base_is_a(&self.as_vtk_object_base(), class_name)
    }

    #[doc(alias = "GetNumberOfGenerationsFromBaseType")]
    fn get_number_of_generations_from_base(&self, base_class: &str) -> i64 {
        ffi::vtk_object_base_get_number_of_generations_from_base(
            &self.as_vtk_object_base(),
            base_class,
        )
    }

    /* TODO this is currently not working yet
    * fn fast_delete(self)
    where
        Self: Sized + Drop,
    {
        let mut sself = self;
        let pin = sself.as_vtk_object_base_mut();
        ffi::vtk_object_base_fast_delete(pin);
        core::mem::drop(sself);
    }*/

    #[doc(alias = "GetReferenceCount")]
    fn get_reference_count(&self) -> i64 {
        ffi::vtk_object_base_get_reference_count(&self.as_vtk_object_base())
    }

    #[doc(alias = "SetReferenceCount")]
    fn set_reference_count(&mut self, count: i64) {
        ffi::vtk_object_base_set_reference_count(self.as_vtk_object_base_mut(), count)
    }

    #[doc(alias = "GetIsInMemkind")]
    fn get_is_in_memkind(&self) -> bool {
        ffi::vtk_object_base_get_is_in_memkind(&self.as_vtk_object_base())
    }

    #[doc(alias = "PrintSelf")]
    fn print_self(&self, indent: u64) -> String {
        ffi::vtk_object_base_print_self(&self.as_vtk_object_base(), indent)
    }

    #[doc(alias = "PrintHeader")]
    fn print_header(&self, indent: u64) -> String {
        ffi::vtk_object_base_print_header(&self.as_vtk_object_base(), indent)
    }

    #[doc(alias = "PrintTrailer")]
    fn print_trailer(&self, indent: u64) -> String {
        ffi::vtk_object_base_print_trailer(&self.as_vtk_object_base(), indent)
    }

    #[cfg(feature = "v094")]
    #[cfg_attr(docsrs, doc(cfg(feature = "v094")))]
    fn uses_garbage_collector(&self) -> bool {
        ffi::vtk_object_base_uses_garbage_collector(&self.as_vtk_object_base())
    }
}
